### DISCUSSION TITLE
Create and train model in BatchLoader when you have another inputs alongside your graph datasets.

### DISCUSSION THREAD

**Discussion by behroozazarkhalili** [danielegrattarola/spektral#231] [Q&A]
Hi.
I am curious to know how to create and train a model in BatchLoader when you have another input alongside graph datasets.
For example, you can create and train protein-ligand interaction problems when your ligand dataset is a set of graphs with their features. Each sample of your protein datasets is a NumPy array of the one-hot encoded sequence of amino acids.

To clarify more, consider the KIBA, DAVIS, BindingDB datasets where each smiles molecules have been represented as a graph (created with deepchem feat module), and each protein has been described as a one-hot encoded amino acid sequence.

How should I inject protein information into my model architecture and train the model when utilizing BatchLoader?
It would be great to add such a model that generalized this excellent package to be applied for more general datasets.
Many thanks for considering my request.

**Comment by danielegrattarola** [comment]
Hi, 

unfortunately what you ask is not directly supported in BatchLoader, but if you are willing to drop the `model.fit` approach then BathLoader is essentially an iterator over the dataset and Graphs are just containers for the adjacency matrix and features. 

You can either write your own training loop where you iterate over the dataset by hand (there's a `bath_generator` util that you can use, if you want) or you can extend BatchLoader to deal with your extra data (you can add the extra data as a special attribute of each graph when you create the dataset).

Unfortunately there is no elegant way to make a Loader that covers all possible use cases, so it's better if Loaders only handle graph data and more advanced use cases are left to the users.

Cheers

**Comment by behroozazarkhalili** [comment]
Hi,
I'm curious to know how I should use the models and layers without using DataLoader. Indeed, my question is whether  I can use layers with my data generator, and if yes, how should the format of the adjacency matrix and features be? For example, should the Adjacency matrix be sparse?

Sorry, but it isn't easy to use your package without clear examples and more concise documentation. I haven't seen any example as to how to use your models with pure TensorFlow datasets and custom generator.

**Comment by danielegrattarola** [comment]
Hi, 

here are examples for node-level and graph-level prediction with custom training loops written in pure tensorflow. 
They still use a Loader, but that's just because a loader is technically a generator.

https://github.com/danielegrattarola/spektral/blob/master/examples/node_prediction/citation_gcn_custom.py
https://github.com/danielegrattarola/spektral/blob/master/examples/graph_prediction/qm9_ecc.py

If you want to write your own loaders, you can. Check out the documentation of each layer to know which inputs they require and in which shape/type (e.g., https://graphneural.network/layers/convolution/#gcnconv).
To understand the different ways that you can create batches of inputs, check out this page: https://graphneural.network/data-modes/

If it is not specified, then the layers will support sparse and dense adjacency matrices. Otherwise, you will see a note in the documentation (e.g., https://graphneural.network/layers/convolution/#edgeconv).

Let me know if you need further assistance.
Cheers
